File: java/cours/src/session2/td/agregation/NatCalculantRecursivement.java
package session2.td.agregation;

public class NatCalculantRecursivement extends NatDeleguantEtat {
	
	public NatCalculantRecursivement(EtatNaturelPur etat) {
		super(etat);
	}

	@Override
	public Nat creerNatAvecEtat(EtatNaturelPur etat) {
		return new NatCalculantRecursivement(etat);
	}

	@Override
	public Nat somme(Nat x) {
		return this.creerSuccesseur(this.predecesseur().somme(x));
	}

	@Override
	public Nat zero() {
		return this.creerZero();
	}

	@Override
	public Nat produit(Nat x) {
		return x.somme(this.predecesseur().produit(x));
	}

	@Override
	public Nat un() {
		return this.creerSuccesseur(this.zero());
	}

	@Override
	public Nat modulo(Nat x) {
		Nat r = this.predecesseur().modulo(x);
		return this.creerSuccesseur(r).equals(x) ? this.creerZero() : this.creerSuccesseur(r);
	}

	@Override
	public Nat div(Nat x) {
		Nat r = this.predecesseur().modulo(x);
		Nat q = this.predecesseur().div(x);
		return this.creerSuccesseur(r).equals(x) ? this.creerSuccesseur(q) : q;
	}
	
	@Override
	public String toString() {
		return "S^" + this.val() + "(0)";
	}
	
	@Override
	public boolean equals(Object obj) {
		if(!(obj instanceof Nat))
			return false;
		Nat x = (Nat)obj;
		if(x.estNul())
			return false;
		return this.predecesseur().equals(x.predecesseur());
	}
}

File: typescript/session2/agregation/NatCalculantRecursivement.ts
import {NatDeleguantEtat} from "./NatDeleguantEtat";
import {EtatNaturelPur} from "./EtatNaturelIPur";
import {Nat} from "./Nat";
import {NatCaculantAvecDesInts} from "./NatCalculantAvecDesInts";
import {NatCalculantAvecDesNombresDecimaux} from "./NatCalculantAvecDesNombresDecimaux";

export class NatCalculantRecursivement extends NatDeleguantEtat {

    constructor(etat : EtatNaturelPur) {
        super(etat);
    }

    creerNatAvecEtat(etat : EtatNaturelPur) : Nat {
        return new NatCalculantRecursivement(etat);
    }

    somme(x : Nat) : Nat {
        return this.creerNatSucc(this.predecesseur().somme(x));
    }

    zero() : Nat {
        return this.creerNatZero();
    }

    produit(x : Nat) : Nat {
        return x.somme(this.predecesseur().produit(x));
    }

    un() : Nat {
        return this.creerNatSucc(this.zero());
    }

    modulo(x : Nat) : Nat {
        let r = this.predecesseur().modulo(x);
        return this.creerNatSucc(r).val() == x.val() ? this.creerNatZero() : this.creerNatSucc(r);
    }

    div(x : Nat) : Nat {
        let r = this.predecesseur().modulo(x);
        let q = this.predecesseur().div(x);
        return this.creerNatSucc(r).val() == x.val() ? this.creerNatSucc(q) : q;
    }

    toString() : string {
        return "S^" + this.val() + "(0)";
    }

    equals(obj : Object) : boolean {
        if(!(NatCalculantRecursivement.instanceOfNat(obj)))
            return false;
        let x = <Nat>obj;
        if(x.estNul())
            return false;
        return this.predecesseur().val() == x.predecesseur().val();
    }

    public static instanceOfNat(nb : Object) : boolean {
        return nb instanceof NatCalculantRecursivement || nb instanceof NatCaculantAvecDesInts || nb instanceof NatCalculantAvecDesNombresDecimaux
    }
}
